catalog/core/catalog-core-api-impl/src/main/java/ddf/catalog/util/impl/SourcePoller.java

```diff
@@ -13,110 +13,285 @@
+    final Map<SourceKey, Source> currentSourceKeyAndSourceMap =
+        currentSources
+            .stream()
+            .collect(
+                Collectors.toMap(
+                    SourceKey::new,
+                    Function.identity(),
+                    (source, anotherSourceWithTheSameSourceKey) -> {
+                      throw new IllegalStateException(
+                          "Found different sources with the same comparable details (id="
+                              + source.getId()
+                              + " for both sources). This means that this SourcePoller might not report the accurate values for the current sources. Confirm that every source has a unique id, and try restarting.");
+                    }));
 
-    LOGGER.debug("Cancelling scheduled polling.");
+    // Remove orphaned SourceKeys or SourceKeys for destroyed Sources from the cache
+    final Set<SourceKey> sourceKeysInTheCache = cache.asMap().keySet();
+    final Set<SourceKey> sourceKeysToRemoveFromTheCache =
+        Sets.difference(sourceKeysInTheCache, currentSourceKeyAndSourceMap.keySet());
+    if (!sourceKeysToRemoveFromTheCache.isEmpty()) {
+      if (LOGGER.isDebugEnabled()) {
+        LOGGER.debug(
+            "Found {} entries to remove from the cache: {}. This is caused by when sources are destroyed or modified in a way that causes their SourceKey to be different since the last poll",
+            sourceKeysToRemoveFromTheCache.size(),
+            StringUtils.join(sourceKeysToRemoveFromTheCache, ", "));
+      }
+      sourceKeysToRemoveFromTheCache
+          .stream()
+          .peek(
+              sourceKey1 -> LOGGER.debug("Removing entry sourceKey {} from the cache", sourceKey1))
+          .forEach(cache::invalidate);
+    } else {
+      LOGGER.trace(
+          "Found no entries to remove from the cache. There has been no sources destroyed or modified in a way that causes their SourceKey to be different since the last poll");
+    }
 
-    runner.shutdown();
+    final int currentSourcesCount = currentSources.size();
 
-    handle.cancel(true);
+    // Loop through each of the currentSources
+    LOGGER.trace(
+        "Starting a process to get the value for each of the {} current sources",
+        currentSourcesCount);
+    final CompletionService<Runnable> completionService =
+        new ExecutorCompletionService<>(pollTimeoutWatcherThreadPool);
+    for (Map.Entry<SourceKey, Source> sourceKeySourceEntry :
+        currentSourceKeyAndSourceMap.entrySet()) {
+      final SourceKey sourceKey = sourceKeySourceEntry.getKey();
+      final Source source = sourceKeySourceEntry.getValue();
 
-    scheduler.shutdownNow();
+      final Future<T> future;
+      try {
+        future = pollThreadPool.submit(() -> getValue(source));
+      } catch (RejectedExecutionException e) {
+        // TODO
+        throw new IllegalStateException("TODO");
+      }
+
+      final Callable<Runnable> callable =
```

Comment:
✏️ I'd recommend using a static inner class here instead of inlining the `Callable` as a lambda. Either that, or extract the code that submits the Callables to the `CompletionService` to a separate private method.